### PR TITLE
Outgoing processing

### PR DESCRIPTION
**Pull request by remvst**
Adds middleware for message post processing. Should be useful for tracking outgoing messages, debugging...

- [ ] Bump version number

@mproberts what do you think?

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -655,6 +664,27 @@ class Bot {
...
 
+    /**
+     *  Runs the outgoing handlers on the specified message.
+     *  @return {promise.<object>}
+     */
+    postProcessMessage(message) {
+        return new Promise((fulfill, reject) => {
+            const outgoingStack = this.outgoingStack.slice(0);
+
+            function runNextHander() {
```
Comment (davidbistolas) [inline index.js]: typo? runNextHandler?

Comment (remvst) [inline index.js]: whoops

**Comment by remvst** [comment]
@davidbistolas fixed

**Comment by arunthampi** [comment]
Nice! LGTM 🙌

**Comment by arunthampi** [comment]
would love to get this merged asap so we can add support for it in our middleware

File: test/test-message-routing.js [hunk 2]
```diff
@@ -802,4 +802,88 @@ describe('Reply handling', () => {
...
+            username: BOT_USERNAME,
+            apiKey: BOT_API_KEY,
+            skipSignatureCheck: true
+        });
+
+        bot.use((incoming, next) => {
+            incoming.markRead();
+        });
+
+        bot.outgoing((outgoing, next) => {
```
Comment (Eagerod) [inline test/test-message-routing.js]: @remvst Commenting out everything in this test relating to outgoing messages doesn't seem to make tests fail. It doesn't feel like this test asserts any behaviour in the code in the rest of the PR. (Removing the feature entirely still has this test passing)

**Comment by remvst** [comment]
@Eagerod test should be fixed now, good catch 👍

**Comment by Eagerod** [comment]
@remvst 👍  🎉

**Comment by arunthampi** [comment]
@remvst @Eagerod thanks again for doing this super quickly. We're using this now in our analytics middleware: https://github.com/botmetrics/botmetrics-kik-middleware

**Pull request merged**